drivers: display: Add ist3931 driver

shenyi97 opened the pull request:
Add the driver  of diplay controller ist3931, i2c/spi 128x64 monocolor display driver. Only the i2c mode driver of the controller is implemented and tested.

decsny commented on dts/bindings/display/ist,ist3931.yaml:
please, make the property names more than two or three characters and better descriptions, this is not acceptable quality

decsny commented on dts/bindings/vendor-prefixes.txt:
```diff
@@ -320,6 +320,7 @@ isee	ISEE 2007 S.L.
 isentek	Isentek Inc.
 isil	Intersil
 issi	Integrated Silicon Solutions Inc.
+ist Intergrated Solutions Technology Inc.
```
I think this is spelled wrong

danieldegrasse commented on drivers/display/display_ist3931.c:
```diff
@@ -0,0 +1,318 @@
+
+struct ist3931_config {
+	union ist3931_bus bus;
+	ist3931_bus_ready_fn bus_ready;
```
This is a good idea- it will make abstracting this driver easier in the future. However, it doesn't seem like you actually use this pointer (just call `ist3931_bus_ready` directly). I am ok with either implementation method (IE using a function pointer or a generically named function), but please remove the code related to the method that isn't used.

At a broader level, please make sure you're consistent with whichever method you choose to abstract the bus. IE if you choose to use generically named functions (`ist3931_bus_ready`, `ist3931_write_bus`) be consistent there, or define function pointers for the write bus and bus ready functions

shenyi97 replied:
Thanks, I will fix the problem.

danieldegrasse replied:
This issue still needs to be fixed

danieldegrasse commented on drivers/display/display_ist3931.c:
```diff
@@ -0,0 +1,318 @@
+			return -EIO;
+		}
+
+		if (ist3931_init_device(dev)) {
```
This feels incorrect- we only setup the device if a reset GPIO is present? If one was not present, what code would setup the display driver during boot?

shenyi97 replied:
This display controller need a reset gpio to operate normally. Do I need to add some extra code to handle the miss of a reset gpio?

danieldegrasse replied:
No, instead you should make the `reset-gpios` property in the devicetree binding be `required: true`, and then you should check to make sure the `reset_gpio` is ready at boot (look at `gpio_is_ready_dt`). If the reset GPIO device isn't ready, the initialization function should return an error.

shenyi97 replied:
Thanks, I will make the `reset-gpios` property be `required:true` in dts binding. The status of reset gpio is already check in line 204 `if (!gpio_is_ready_dt(&config->reset_gpio)) {
	LOG_ERR("Reset GPIO device not ready");
	return -ENODEV;
}`.

danieldegrasse replied:
This still needs to be addressed- `reset-gpios` is now `required:true`, but the `if (config->reset_gpio.port != NULL) {` block here wrapping this code should be dropped. We also should propagate the return codes from functions when they fail. I believe this code block should then look like so:

```
int ret;

if (!gpio_is_ready_dt(&config->reset_gpio)) {
	LOG_ERR("Reset GPIO device not ready");
	return -ENODEV;
}
ret = gpio_pin_configure_dt(&config->reset_gpio, GPIO_OUTPUT_INACTIVE);
if (ret) {
	LOG_ERR("Couldn't configure reset pin");
	return ret;
}

ret = ist3931_init_device(dev)
if (ret) {
	LOG_ERR("Failed to initialize device!");
	return ret;
}
```

danieldegrasse commented on drivers/display/display_ist3931.c:
```diff
@@ -0,0 +1,318 @@
+	return 0;
+}
+
+static int ist3931_blanking_on(const struct device *dev)
```
If I understand this correctly, the implementation here is going to clear the display data, which is not what the blanking API should do. When `display_blanking_off` is called, the data that was on the display should still be present, and `display_blanking_on` should not clear the framebuffer.

Is there some command that could be used to blank the display? The implementations of `blanking_on` often simply disable a backlight control.

shenyi97 replied:
Thanks, I misunderstood the `display_blanking_off `api. I will fix it.

danieldegrasse commented on drivers/display/display_ist3931.c:
```diff
@@ -0,0 +1,318 @@
+	return 0;
+}
+
+static int ist3931_blanking_off(const struct device *dev)
```
See above comment on the `blanking_off` implementation- this is not how the blanking API should function

danieldegrasse commented on drivers/display/display_ist3931.h:
```diff
@@ -0,0 +1,38 @@
+#define IST3931_CMD_POWER_CONTROL          0x2c
+#define IST3931_CMD_BIAS                   0x30
+#define IST3931_CMD_CT                     0xb1
+#define IST3931_CMD_Frame_Control          0xb2
```
Nit- these command definitions should be all CAPS

danieldegrasse commented on drivers/display/display_ist3931.h:
```diff
@@ -0,0 +1,38 @@
+#define IST3931_CMD_DELAY   10
+#define IST3931_RAM_WIDTH   144
+#define IST3931_RAM_HEIGHT  65
+#endif /* _ZEPHYR_DRIVERS_DISPLAY_IST3931_H_ */
```
Add newline here

danieldegrasse commented on dts/bindings/vendor-prefixes.txt:
The commit adding this vendor prefix should be first in the commit series

danieldegrasse commented on dts/bindings/display/ist,ist3931.yaml:
Commit with this binding should be present second in commit series

danieldegrasse reviewed (commented):
Thanks for the driver- can you add an entry with this display into the `tests/drivers/build_all/display` application overlay, so this driver is built by CI?

shenyi97 commented on drivers/display/CMakeLists.txt:
```diff
@@ -12,6 +12,8 @@ zephyr_library_sources_ifdef(CONFIG_ILI9340		display_ili9340.c)
 zephyr_library_sources_ifdef(CONFIG_ILI9341		display_ili9341.c)
 zephyr_library_sources_ifdef(CONFIG_ILI9342C	display_ili9342c.c)
 zephyr_library_sources_ifdef(CONFIG_ILI9488		display_ili9488.c)
+zephyr_library_sources_ifdef(CONFIG_ILI9486		display_ili9486.c)
```
It is a mistake from copilot, I will delete this line.

shenyi97 commented on drivers/display/display_ist3931.c:
```diff
@@ -0,0 +1,318 @@
+};
+
+struct ist3931_config {
+	union ist3931_bus bus;
```
This display controller do support i2c and spi 3 wire, but only the i2c mode driver is implemented. I will take your advice.

shenyi97 commented on drivers/display/display_ist3931.c:
```diff
@@ -0,0 +1,318 @@
+	uint8_t y_offset;
+};
+
+static int ist3931_write_bus_i2c(const struct device *dev, uint8_t *buf, bool command,
```
Thanks, it will be renamed to `ist3931_write_bus`

shenyi97 commented on drivers/display/display_ist3931.c:
```diff
@@ -0,0 +1,318 @@
+	int ret;
+	const struct ist3931_config *config = dev->config;
+	const uint8_t command_byte = command ? IST3931_CMD_BYTE : IST3931_DATA_BYTE;
+	uint8_t *i2c_write_buf = k_malloc(num_bytes * 2);
```
Thanks. I tried on a esp32s3 board but failed, so I checked `i2c_transfer `api and found this.
> In some cases drivers will generate separate transactions for each message fragment, with or without presence of [I2C_MSG_RESTART] in [flags].

This display controller need an i2c sequence like `START | ADDRESS BYTE|  COMMAND BYTE | ACK| DATA BYTE | ACK| COMMAND BYTE | ACK| DATA BYTE | STOP`. I think it means I can't use `i2c_transfer` to transfer the command byte and then use `i2c_transfer` to transfer the data byte, since the I2C_MSG_RESTART will interrupt the i2c sequence.

danieldegrasse replied:
Hmm, yeah- reading the API, this does seem to be the case. I see now that a control byte needs to be sent with *every* data byte, is that correct? I do think we need an intermediate buffer for this based on the way the I2C API is described. Perhaps we can statically allocate this buffer to be up to `IST3931_RAM_WIDTH/4` bytes, since it seems like that would be the largest transfer ever made to this display (from the `display_write` implementation, unless I have missed something)

shenyi97 commented:
@danieldegrasse @decsny Is there anything I can do to let this PR mergerd?

danieldegrasse reviewed (commented):
Still need to clean up the definition of the bus ready and transfer functions, see my comment response

danieldegrasse commented:
> @danieldegrasse @decsny Is there anything I can do to let this PR mergerd?

@shenyi97 just a ping here- Please take a look at my review comment if you'd like to move this driver forwards: https://github.com/zephyrproject-rtos/zephyr/pull/72116#discussion_r1598656982

kartben commented:
@shenyi97 will you be able to come back to this pull request?

shenyi97 commented:
> @shenyi97 will you be able to come back to this pull request?

yes, I will come back soon

Commit:
dts: bindings: add vendor prefix of istech

Add vendor prefix for Integrated Solutions Technology Inc.

Signed-off-by: Shen Xuyang <shenxuyang@shlinyuantech.com>

Commit:
dts: bindings: Add dts binding of ist3931

Add dts binding for istech,ist3931

Signed-off-by: Shen Xuyang <shenxuyang@shlinyuantech.com>

danieldegrasse commented on drivers/display/display_ist3931.c:
```diff
@@ -0,0 +1,289 @@
+#define IST3931_INIT(inst)                                                                         \
+	static const struct ist3931_config ist3931_config_##inst = {                               \
+		.bus = I2C_DT_SPEC_INST_GET(inst),                                                 \
+		.reset_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, reset_gpios, {}),                     \
```
```suggestion
		.reset_gpio = GPIO_DT_SPEC_INST_GET(inst, reset_gpios),                     \
```

(the `reset-gpios` property is `required: true`)

danieldegrasse commented on drivers/display/display_ist3931.h:
```diff
@@ -0,0 +1,38 @@
+#ifndef IST3931_DISPLAY_DRIVER_H__
+#define IST3931_DISPLAY_DRIVER_H__
+
+#include <zephyr/kernel.h>
```
Why are including the kernel header here? If the driver needs this header, include it in the driver C file.

kartben commented on drivers/display/display_ist3931.c:
```diff
@@ -0,0 +1,289 @@
+	if (ret < 0) {
+		return ret;
+	}
+	ret = ist3931_driver_set_ax(dev, y);
```
```suggestion
	ret = ist3931_driver_set_ax(dev, x);
```

shenyi97 replied:
Thanks, this mistake will be corrected

Commit:
driver: display: Add driver for diplay controller ist3931

Add the driver of diplay controller ist3931, i2c/spi
128x64 monocolor display driver. Only the i2c mode
driver of the controller is implemented and tested.

Signed-off-by: Shen Xuyang <shenxuyang@shlinyuantech.com>

Commit:
tests: drivers: Add overlay for test of IST3931 display driver

Add the overlay for the test of IST3931 display driver

Signed-off-by: Shen Xuyang <shenxuyang@shlinyuantech.com>
